proposals/0000-asset-primitive.md

```diff
@@ -0,0 +1,233 @@
+
+### Goals
+
+This proposal outlines an **asset primitive** that accomplishes the following goals:
+
+- Works for Fonts, Icons, and Images.
+- Can reasonably extend to other asset types in the future.
+- Can reasonably extend to support image optimizations in the future.
+- Works with SSR, where all assets must be known at build-time, with 100% accuracy.
+- Works across frameworks (`.jsx`, `.svelte`, `.vue`). Not an `.astro`-only solution.
+- Introduces an acceptable amount of boilerplate (as little as possible, or none).
+- Can power custom 3rd-party integrations and components.
+- Can power future, higher-level Astro core features.
+
+# Detailed design
+
+## The Asset Primitive
+
+```js
+import imageBook from '../assets/book.png';
+  // imageBook is type: ImageAsset (.png|.jpg|.jpeg|.webp)
+  // {url, width, height, blurHashUrl}
+import iconCookie from '../assets/cookie.svg';
+  // iconCookieis type: SvgAsset (.svg)
+  // {width, height, content}
+import robotoFont400 from '../assets/fonts/roboto-slab-v22-latin-regular.woff2';
+  // robotoFont400 is type: string
+  // [If no custom asset is defined, Astro will default to including the asset
+  // in your build and returning a single URL as the result. This is Vite default behavior.]
+```
+
+- The asset primitive is imported via ESM.
+- Astro will define two custom `Asset` types: `ImageAsset` and `SvgAsset`
+- All other assets will just return a URL path string to the asset, and include that asset in the final build.
+- Works in `.astro` and all component files (`.jsx`, `.svelte`, `.vue`, etc)
+- Builds on top of the accepted [`local:src`](https://github.com/withastro/rfcs/blob/assets-rfc/proposals/0011-relative-url-scheme.md) RFC.
+
+## The Asset Components
+
+```astro
+---
+import imageBook from '../assets/book.png';
+import iconCookie from '../assets/cookie.svg';
+import robotoFont400 from '../assets/fonts/roboto-slab-v22-latin-regular.woff2';
+import {Image, Icon, Font} from 'astro/components';
+---
+<html>
+<head>
+  <Font use={robotoFont400} family="Roboto Slab" weight={400} style="normal" />
+</head>
+<body>
+  <Icon use={IconCookie} />
+  <Image use={imageBook} />
+</body>
+</html>
+```
+
+- Astro will ship `Font`, `Icon`, and `Image` components that can take these asset types, and properly use them on the page:
+  - `Font`: Creates a `<style>` tag with a single `@font-family` definition
+  - `Image`: Creates a `<img>` tag with `src=` the asset URL
+  - `Icon`: renders the given `svg` content to HTML.
+- In the future, these can be extended to add new features. For example, the Image component can be extended to support image optimizations and automatic [blurhash](https://blurha.sh/) support.
```

Comment:
I think this is the most important part of figuring out assets. I don't think these components *can* be extended to do those kinds of things, as there are missing primitives around transforming files either during the build or at runtime.

If you want to make this a future proposal I think that's fine, but in that case I don't think the builtin components should be added until that part is ready. A 3rd party should have the same ability to do transformations as a builtin component and I think shipping builtins before that sends a signal to the community that the builtins are going to get special APIs that they won't have access to.

We made this mistake with the Markdown component (giving special APIs) and haven't yet been able to undo it. Once something is implemented high-level there is less of an incentive to ever pull out the primitives, so I'd be disappointed if we did this again with asset components.

Reply:
I think I'm aligned on this. The built-in components were included to make it clear how you would use these assets, but I can illustrate this without actually proposing that we build them before we are ready.